FurkannOzbek/application-hub
# Implement Token-Based Authentication and Logout Functionality

**Description**
This PR introduces a token-based authentication mechanism for the application, enhancing security by ensuring that API access is restricted to authorized users only. It also integrates token invalidation upon user logout to prevent unauthorized reuse of tokens. Additionally, it modifies the user context in the frontend to properly handle authentication states and logout.

- **Problem Solved:** Implements proper token-based authentication and logout functionality to address security concerns and streamline the user session management process.
- **Impact:** Ensures secure API access, adds middleware to validate tokens, enables token invalidation, and aligns frontend authentication flows with backend changes.

**Changes Made**
        **Frontend:**

-  Updated Authentication.js to handle logout by calling the new /logout API endpoint and managing local storage for user session data.
-  Introduced makeLogoutApiCall.js utility to send logout requests to the backend.
-  Adjusted the frontend authentication flow to align with backend token-based authentication.

      **Backend:**

- Added middleware/token_authentication.js to verify JWT tokens and block invalidated tokens.
- Created a new routers/logout.js to handle token invalidation during logout by storing token IDs in the database.
- Updated routers/user.js to use the new token authentication middleware.
- Enhanced auth.js to generate JWT tokens with jti for tracking and added functionality to invalidate tokens.
  
**General:**
-    Updated package.json to include necessary dependencies (uuid).

**Testing Steps**
    **Frontend:**
1. Login to the application and verify that a valid token is stored in local storage.
2. Trigger logout and confirm that:
- The token is removed from local storage.
- The user is redirected to the login page.

     **Backend:**

1. Test the /login endpoint to verify the token generation.
2. Test the /logout endpoint with valid and invalid tokens to ensure proper invalidation.
